agent/consul/acl_replication.go

```diff
@@ -1,347 +1,556 @@
+	}
+
+	if len(deletions) > 0 {
+		s.logger.Printf("[DEBUG] acl: policy replication - performing deletions")
+
+		exit, err := s.deleteLocalACLPolicies(deletions, stopCh)
+		if exit {
+			return 0, true, nil
+		}
+		if err != nil {
+			return 0, false, fmt.Errorf("failed to delete local ACL policies: %v", err)
 		}
+		s.logger.Printf("[DEBUG] acl: policy replication - finished deletions")
+	}
 
-		// Do a smooth rate limit to wait out the min time allowed for
-		// each op. If this op took longer than the min, then the sleep
-		// time will be negative and we will just move on.
-		elapsed := time.Since(start)
-		time.Sleep(minTimePerOp - elapsed)
+	if len(updates) > 0 {
+		s.logger.Printf("[DEBUG] acl: policy replication - performing updates")
+		exit, err := s.updateLocalACLPolicies(policies.Policies, stopCh)
+		if exit {
+			return 0, true, nil
+		}
+		if err != nil {
+			return 0, false, fmt.Errorf("failed to update local ACL policies: %v", err)
+		}
+		s.logger.Printf("[DEBUG] acl: policy replication - finished updates")
 	}
-	return nil
+
+	// Return the index we got back from the remote side, since we've synced
+	// up with the remote state as of that index.
+	return remote.QueryMeta.Index, false, nil
 }
 
-// replicateACLs is a runs one pass of the algorithm for replicating ACLs from
-// a remote ACL datacenter to local state. If there's any error, this will return
-// 0 for the lastRemoteIndex, which will cause us to immediately do a full sync
-// next time.
-func (s *Server) replicateACLs(lastRemoteIndex uint64) (uint64, error) {
-	remote, err := s.fetchRemoteACLs(lastRemoteIndex)
+func (s *Server) replicateACLTokens(lastRemoteIndex uint64, stopCh <-chan struct{}) (uint64, bool, error) {
+	remote, err := s.fetchACLTokens(lastRemoteIndex)
 	if err != nil {
-		return 0, fmt.Errorf("failed to retrieve remote ACLs: %v", err)
+		return 0, false, fmt.Errorf("failed to retrieve remote ACL tokens: %v", err)
 	}
 
-	// This will be pretty common because we will be blocking for a long time
-	// and may have lost leadership, so lets control the message here instead
-	// of returning deeper error messages from from Raft.
-	if !s.IsLeader() {
-		return 0, fmt.Errorf("no longer cluster leader")
+	s.logger.Printf("[DEBUG] acl: finished fetching remote tokens: %d", len(remote.Tokens))
+
+	// Need to check if we should be stopping. This will be common as the fetching process is a blocking
+	// RPC which could have been hanging around for a long time and during that time leadership could
+	// have been lost.
+	select {
+	case <-stopCh:
+		return 0, true, nil
+	default:
+		// do nothing
 	}
 
 	// Measure everything after the remote query, which can block for long
 	// periods of time. This metric is a good measure of how expensive the
 	// replication process is.
-	defer metrics.MeasureSince([]string{"leader", "replicateACLs"}, time.Now())
+	defer metrics.MeasureSince([]string{"leader", "replication", "acl", "token", "apply"}, time.Now())
```

Comment:
Is this a new metric? The policies one is certainly, do you have updating the telemetry docs on your TODO list?

Reply:
Added a todo for this.